### ISSUE TITLE
Backup Data associated with a particular Api Key

### ISSUE THREAD

**Issue by priyankachaturvedi30** [Visual-Regression-Tracker/Visual-Regression-Tracker#184]
@pashidlos : Since the setup is through the dockerized container, each time I run docker-compose up would generate a new API key for me. What If I want to retain the data(images/projects/builds) associated with a particular API key? Or maybe something through which I could get the data(associated with old API key) backed up to import in a container with a new API key? Is there feasibility to do something on similar lines?

**Comment by pashidlos** [comment]
@priyankachaturvedi30 could you what steps when API key is re-generated?
it should be generated only once per user registration 
whenever  you `docker-compose up` it should reuse before existed containers and all data should be persisted
every time on start up script is checking if DB contains any user, if not it generates a default one 
please, check if your are removing container/DB at some point

**Comment by priyankachaturvedi30** [comment]
I do not understand when you say user registration? Yes, it should reuse the existing containers and all data should be persisted but I got the following error when I tried it - 
![image](https://user-images.githubusercontent.com/50797731/105045379-dc6dbd00-5a8d-11eb-8a72-83b0e238fb66.png)

After that when I executed the docker-compose up, it re-generated the API key for me. Also, please note that I have not created any user until now, just using the default one only.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 could you try clearing all docket containers and volumes
And try setup from scratch
In case you have DB data cleared after docket stop/up let me know the steps

**Comment by priyankachaturvedi30** [comment]
Sure @pashidlos. Let me give that a try and get back to you.

**Comment by priyankachaturvedi30** [comment]
@pashidlos : I restarted my docker after closing the existing docker instance and it showed that my docker processes are running 
![image](https://user-images.githubusercontent.com/50797731/105170032-611a1300-5b42-11eb-82ae-1094d4b7f11b.png) However, when I checked the UI, I got no data related to my previous builds. Also, after that when I did a docker-compose up I created new API key for me again and my data got lost :( Also, just thinking, in case my docker instance gets crashed I am skeptical of restoring my data.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 are you using docker toolbox?
according to this page it was deprecated 
https://docs.docker.com/docker-for-windows/docker-toolbox/

it looks like it re-create containers instead of starting them 
could you share `docker-compose` logs ?

**Comment by priyankachaturvedi30** [comment]
Yes @pashidlos I am using Docker Toolbox. But what I see I the logs is that some specific table is missing : **Invalid `prisma.user.findUnique()` invocation: The table `public.User` does not exist in the current database.** Attaching logs for docker-compose for your further reference.
[myDockerCompose.log](https://github.com/Visual-Regression-Tracker/Visual-Regression-Tracker/files/5847425/myDockerCompose.log)

**Comment by pashidlos** [comment]
@priyankachaturvedi30 logs seems corrupted 
could you:
- switch to recent docker for windows https://docs.docker.com/docker-for-windows/
- remove all containers/images/volumes
- start over again 

not sure what happened to make this table disappear unless it's due to docker toolbox issue

**Comment by priyankachaturvedi30** [comment]
Sure @pashidlos : I have cleaned all the containers and unistalled docker toolbox. Have setup the appilcation with the help of Docker Desktop. I'll verify and update you incase I still have this issue.

**Comment by priyankachaturvedi30** [comment]
@pashidlos  One more thing, One of my containers from other processes was also using port 8080, so I changed the port for env and compose file in case of VRT to 8008, but for some reason, I am not able to access the application on 8008 port. Any thought why this is happening?

**Comment by pashidlos** [comment]
@priyankachaturvedi30 changing port require update only in `.env` file in two places 
no changes in `docker-compose.yml` required 
have you restarted docker after?
share docker-compose logs

**Comment by priyankachaturvedi30** [comment]
@pashidlos : One more concern related to data backup here - There can be a case where my docker-machine got crashed or maybe my containers got killed/terminated due to some reason. Considering that my tests are running on a daily basis, it would be a huge task to re-create those baselines ?  What If I want to retain the data(images/projects/builds) associated with a particular API key ?

**Comment by pashidlos** [comment]
@priyankachaturvedi30 your data is stored out of container inside `/imageUploads` and `/db` folders
you could move this folders and recreate everything from them

**Comment by priyankachaturvedi30** [comment]
Great ! Thanks for the info. Let me try this out once :)

**Comment by priyankachaturvedi30** [comment]
@pashidlos I can see my data present inside /imageUploads folder, but when i am doing a docker-compose up its not picking the old baselines/projects. Also, for some reason I have observed that a new API key is also generated. Can you suggest me a way to revert back to the older api key with the data. (Considering i have all the screenshots present in imageUploads folder)

**Comment by pashidlos** [comment]
@priyankachaturvedi30 it could happen after changing DB name in `.env` file (new db gonna be created)
you could check if there are any other DB created by starting postgress container and connecting to using something like https://www.pgadmin.org/
and you'll see the list of already existing DBs 
in order to restore the data you need to use previous DB name in config 
![image](https://user-images.githubusercontent.com/5182956/110310228-fb5aeb00-800a-11eb-9153-2fac693e3c6b.png)

**Comment by priyankachaturvedi30** [comment]
@pashidlos I verified that as well. Nothing being changed in .env file as well. Also, there is only 1 db present 
![image](https://user-images.githubusercontent.com/50797731/110313994-484fb900-802d-11eb-8c08-cee6d39ae799.png).

I can very well see all the screenshots in my imageFolders but still not able to restore data :(

**Comment by pashidlos** [comment]
@priyankachaturvedi30 together with `imageUploads` there should be `db` folder that contains DB related data 
to validate that DB data is restored you could clean docker data by `docker system prune -a` and start containers again `docker-compose up`
all data should be restored if it's present in `db` folder
you could try searching for that folder in case it was moved/removed and then use it to restore the data

**Comment by priyankachaturvedi30** [comment]
yep. I see the db folder as well. I cleaned the docker data and started the container but still no data got restored. I can however see a few error . Attached logs for 
[output.log](https://github.com/Visual-Regression-Tracker/Visual-Regression-Tracker/files/6101237/output.log)
reference.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 there are errors about not logged in user getting auth error, login with correct creds 
in logs I see that initially data was seeded
next time the project started I could see that DB got restored 
```
[32mpostgres_1   |[0m PostgreSQL Database directory appears to contain a database; Skipping initialization
```
only the data that in `db` folder will be restored

**Comment by priyankachaturvedi30** [comment]
@pashidlos  : I have been playing around with different use cases to ensure how the system behaves. Just one more concern here - If the existing containers are removed can i still get backup restored somehow ? Following your previous comment which says we can do so by using imageUploads and db folders - 
![image](https://user-images.githubusercontent.com/50797731/110466901-59b3c680-80fc-11eb-9cc8-3f86adfe0ffd.png)

Can you explain more on the steps as to how can I achieve this? (moving folders and recreating )

**Comment by pashidlos** [comment]
@priyankachaturvedi30 you could move those folders elsewhere 
put them together with `docker-compose.yml` and new containers will be recreated using data out of those folders

**Comment by priyankachaturvedi30** [comment]
So @pashidlos : I think i got the issue why i wasnt able to retain data at some instances.  So following the docker-compose file -
![image](https://user-images.githubusercontent.com/50797731/110931966-de971e00-8350-11eb-9e7e-863e7c8128d9.png)
The 2 volumes were not created for me when i was trying to do docker volume ls. (Not sure why :( ) I then explicity did a volume mount by adding volumes in compose file - 
![image](https://user-images.githubusercontent.com/50797731/110932123-130ada00-8351-11eb-8585-c233be0921ea.png)
and yes, things worked. Infact, now even if my containers are explicity killed/removed i can retain my previous data. Just doing some testing around it to verify the same.

**Comment by pashidlos** [comment]
@priyankachaturvedi30 could you confirm this is the fix?
are you using docker toolbox or docker for windows?
there is an issue for that where people sharing some ideas https://github.com/docker/compose/issues/4303

you could try changing volume to something like (having needed folders in c:/vrt)
- //c/vrt/imageUploads:/imageUploads
- //c/vrt/db:/var/lib/postgresql/data

**Comment by priyankachaturvedi30** [comment]
ok, so I am using Docker for windows. I guess I tried using the following as well but it didn't worked out as expected.
![image](https://user-images.githubusercontent.com/50797731/111108369-bd1b7980-857e-11eb-9391-744656a1630f.png)

So, from my understanding, there are two ways we can achieve it. through bind mounts and docker volume. What was specified in the original docker compose was using bind mounts (which for some reason was not working for me) However, when i used docker volume it worked. I will be doing some more tests and confirm you with the findings @pashidlos

**Comment by priyankachaturvedi30** [comment]
@pashidlos : Looks like there is an issue with the behavior of VRT which is different when we restart it using docker-compose down/up. After doing docker-compose down, In case of windows, it recreates the containers which restoring backup/data while incase of Linux the containers are unable to retain/restore backup. Can you confirm if you have similar observation?

**Comment by pashidlos** [comment]
@priyankachaturvedi30 I'm using docker-compose on macOs and it works
when I upgrade to newest version all the data is getting restored

**Comment by priyankachaturvedi30** [comment]
@pashidlos : What version of docker are you using?

**Comment by pashidlos** [comment]
@priyankachaturvedi30 
docker version
```
Client: Docker Engine - Community
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        2291f61
 Built:             Mon Dec 28 16:17:32 2020
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true
```
docker-compose version
```
docker-compose version 1.27.4, build 40524192
docker-py version: 4.3.1
CPython version: 3.7.7
OpenSSL version: OpenSSL 1.1.0l  10 Sep 2019
```

**Comment by priyankachaturvedi30** [comment]
@pashidlos : The issue has been resolved. The problem was in case of docker-compose down on Linux the containers were getting removed and no data was getting restored due to following reasons.
1. As defined in docker documentation - https://docs.docker.com/compose/reference/down/
**Networks and volumes defined as external are never removed.** -- So I externalized the volume.
2. Since my docker was running 2 separate Postgres (1 for vrt and 1 for another application ), I changed one of the postgres volume as 
![image](https://user-images.githubusercontent.com/50797731/111426058-0fdd6880-871a-11eb-8594-8b26fa91a8fc.png)

Looks like postgres mandatorily requires the volumes for its container in **/var/lib/postgresql/data** location only.

Applying both above changes it worked for me :)
